source/common/upstream/load_balancer_impl.cc

```diff
@@ -20,61 +35,83 @@ static const std::string RuntimePanicThreshold = "upstream.healthy_panic_thresho
 LoadBalancerBase::LoadBalancerBase(const PrioritySet& priority_set,
                                    const PrioritySet* local_priority_set, ClusterStats& stats,
                                    Runtime::Loader& runtime, Runtime::RandomGenerator& random)
-    : stats_(stats), runtime_(runtime), random_(random),
-      host_set_(*priority_set.hostSetsPerPriority()[0]),
-      local_host_set_(local_priority_set ? local_priority_set->hostSetsPerPriority()[0].get()
-                                         : nullptr) {
-  if (local_host_set_) {
-    host_set_.addMemberUpdateCb([this](uint32_t, const std::vector<HostSharedPtr>&,
-                                       const std::vector<HostSharedPtr>&) -> void {
-      regenerateLocalityRoutingStructures();
-    });
-    local_host_set_member_update_cb_handle_ =
-        local_host_set_->addMemberUpdateCb([this](uint32_t, const std::vector<HostSharedPtr>&,
-                                                  const std::vector<HostSharedPtr>&) -> void {
-          regenerateLocalityRoutingStructures();
+    : stats_(stats), runtime_(runtime), random_(random), priority_set_(priority_set),
+      best_available_host_set_(bestAvailable(&priority_set)),
+      local_priority_set_(local_priority_set) {
+  resizePerPriorityState();
+  priority_set_.addMemberUpdateCb([this](uint32_t priority, const std::vector<HostSharedPtr>&,
+                                         const std::vector<HostSharedPtr>&) -> void {
+    // Update the host set to use for picking, based on the new state.
+    best_available_host_set_ = bestAvailable(&priority_set_);
+    // Make sure per_priority_state_ is as large as priority_set_.hostSetsPerPriority()
+    resizePerPriorityState();
+    // If there's a local priority set, regenerate all routing based on a potential size change to
+    // the hosts routed to.
+    if (local_priority_set_) {
+      regenerateLocalityRoutingStructures(priority);
+    }
+  });
+  if (local_priority_set_) {
+    // Multiple priorities are unsupported for local priority sets.
+    // TODO(alyssawilk) find the right place to reject config with this.
```

Comment:
If multiple priorities aren't supported for the local priority set, i.e. it is always a singleton, could it just be a `HostSet` instead of a `PrioritySet`? I'm guessing there is some advantage in being able to treat it as a `PrioritySet` when working with common code, but just wanted to check (and maybe request a comment be added in some place to explain).

Reply:
We could absolutely have local_host_set here.  I hacked that up a ways back and decided it wasn't much better because
- we have to convert from priority set to host set somewhere (the thread local cluster is just another cluster in the ClusterManagerImpl) and it's about as ugly elsewhere and 
- if we do it in ClusterManagerImpl we have to reinstate "add callback" to the upstream HostSet API (which makes me a bit nervous as generally it's the wrong thing to do to subscribe to updates to a host set).

Alternately [I think] we can make the "local envoy hosts" a host set in ClusterManagerImpl, but then we've moved the problem to config updates, as 99% would go to thread_local_clusters_ but updates to local_cluster_name.value() would be applied directly to the special one-off host set.

I don't have super strong feelings either way so take a look at where we set local_priority_set_ and let me know if you think one of those options is worth the extra code churn.

Reply:
Per offline convo, if we were to actually support multiple priority levels for the local cluster, we would actually need to use *all priorities" when computing the locality routing structures if we assume that all Envoys in the local cluster failover at roughly the same time. To support this, IMO we would likely add additional methods at a PrioritySet to get all hosts or something like that. As such I think what we are doing here is probably good for now. I might add some more comments on what the "real" solution would be if we want to implement that ever. I might also add a TODO/comment about the "real" solution so future readers know all the thinking that went into this.

Also, per comment below, I do have a preference for blocking this at the EDS update layer and just using asserts in the LB code.